spec/controllers/users/totp_setup_controller_spec.rb

```diff
@@ -0,0 +1,88 @@
+require 'rails_helper'
+
+describe Users::TotpSetupController, devise: true do
+  render_views
+
+  describe 'before_actions' do
+    it 'includes confirm_two_factor_authenticated' do
+      expect(subject).to have_filters(
+        :before,
+        :confirm_two_factor_authenticated
+      )
+    end
+  end
+
+  describe '#new' do
+    before do
+      sign_in_as_user
+
+      get :new
+    end
+
+    it 'returns a 200 status code' do
+      expect(response.status).to eq(200)
+    end
+
+    it 'sets otp_secret_key in session' do
+      expect(session[:otp_secret_key]).not_to be_nil
+    end
+
+    it 'presents a QR code to the user' do
+      expect(response.body).to include('QR Code for Authenticator App')
+    end
+  end
+
+  describe '#confirm' do
+    context 'when user presents invalid code' do
+      before do
+        sign_in_as_user
+        get :new
+        patch :confirm, code: 123
+      end
+
+      it 'redirects :new' do
+        expect(response).to redirect_to(users_totp_path)
+      end
+
+      it 'sets flash[:error] message' do
+        expect(flash[:error]).to eq t('upaya.errors.invalid_totp')
+      end
+
+      it 'does no enable totp for the current user' do
+        expect(subject.current_user.totp_enabled?).to be(false)
+      end
+    end
+
+    context 'when user presents correct code' do
+      before do
+        sign_in_as_user
+        get :new
+        expect(subject.current_user.totp_enabled?).to be(false)
+        patch :confirm, code: generate_totp_code(session[:otp_secret_key])
+      end
+
+      it 'redirects to edit_user_registration_path' do
+        expect(response).to redirect_to(edit_user_registration_path)
+      end
+
+      it 'sets flash[:success] message' do
+        expect(flash[:success]).to eq t('upaya.notices.totp_configured')
+      end
+
+      it 'enables totp for the current user' do
+        expect(subject.current_user.totp_enabled?).to be(true)
+      end
+    end
+  end
+
+  describe '#disable' do
+    before do
+      sign_in_as_user
+      get :disable
```

Comment:
Shouldn't this be `delete` instead of `get`? Also, this test probably needs to make sure the user has enabled TOTP first so that we can check that clicking disable changes state from enabled to disabled.